### PR TITLE
[Feature] support decorators to the openapi

### LINKED ISSUE #483: Add support for adding decorators to the OpenAPI spec and documentation UI endpoints
As a user of the APIFlask framework, I would like the ability to apply a decorator to the endpoints used to serve the OpenAPI spec as well as the documentation UI, so that I can execute additional logic before returning back the OpenAPI spec or the documentation UI to a client.

The more specific use case I have for this is to apply some security checks to validate that the client trying to access the documentation is allowed to access the endpoints that serve it.

I have currently achieved this by doing the following, but it's rather ugly - I had to look into the code in  `APIFlask._register_openapi_blueprint` and replicate some of its logic to rewrite the view functions, which Flask advises against doing:

```
def _add_authentication_to_docs(app: APIFlask) -> None:
    if app.spec_path:
        _rewrite_view_function(app, "spec")
    if app.docs_path:
        _rewrite_view_function(app, "docs")
        if app.docs_ui == "swagger-ui" and app.docs_oauth2_redirect_path:
            _rewrite_view_function(app, "swagger_ui_oauth_redirect")


def _rewrite_view_function(app: APIFlask, method_name: str) -> None:
    # Flask warns you not to do this, but we don't really have a choice...
    app.view_functions[f"openapi.{method_name}"] = authenticate_apidocs()(app.view_functions[f"openapi.{method_name}"])
```

APIFairy provides this functionality via the `APIFAIRY_APISPEC_DECORATORS` and `APIFAIRY_UI_DECORATORS` config variables, which is much more straightforward:

```
app.config["APIFAIRY_APISPEC_DECORATORS"] = [authenticate_apidocs()]
app.config["APIFAIRY_UI_DECORATORS"] = [authenticate_apidocs()]
```

I am hoping that this framework might consider adopting similar functionality.

### PR DESCRIPTION
**Pull request by FarmerChillax**
fixes #483 

Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the `docs` folder and in code docstring.
- [x] Add an entry in `CHANGES.md` summarizing the change and linking to the issue.
- [x] Add `*Version changed*` or `*Version added*` note in any relevant docs and docstring.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

File: examples/openapi/custom_decorators/app.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+users = {
+    'userA': generate_password_hash('foo'),
+    'userB': generate_password_hash('bar'),
+}
+
+
+# if you wanna apply an authentication or other decorator
+# to the endpoints used to server the OpenAPI spec
+# as well as the documentation UI, you can used like this.
```
Comment (wlach) [inline examples/openapi/custom_decorators/app.py:17]:
```suggestion
# if you want to apply an authentication or other decorator
# to the endpoints used to serve the OpenAPI spec
# as well as the documentation UI, use the
# `APISPEC_DECORATORS` and `DOCS_DECORATORS`
# configuration options
```

**Review by wlach — COMMENTED** [review]
(just a drive-by suggestion, not the maintainer) This looks great! I wanted this feature and had to resort to various hacks involving using APIFlask internals to get it to work. This is 100x better.

File: examples/openapi/custom_decorators/app.py (hunk 1 above)
Comment (FarmerChillax) [reply examples/openapi/custom_decorators/app.py:17]: this looks better

**Comment by tsokalski** [comment]
This looks good to me too!

File: src/apiflask/app.py [hunk 2]
```diff
@@ -1193,3 +1201,20 @@ def _generate_spec(self) -> APISpec:
...
+        """inject the decorators in runtime.
+
+        Args:
+            config_name (str): inject config name
+        """
+        def wrapper(func):
+            @wraps(func)
+            def inject_decorators(*args, **kwargs):
+                spec_func = func
+                apispec_decorators = self.config.get(config_name, [])
```
Comment (greyli) [inline src/apiflask/app.py:1215]: I would recommend using a general variable name here (i.e. `decorators`). The decorators could be fetched from three different configs (not just `APISPEC_DECORATORS`).

**Review by greyli — CHANGES_REQUESTED** [review]

File: docs/configuration.md [hunk 3]
```diff
@@ -576,6 +599,18 @@ app.config['SPEC_PROCESSOR_PASS_OBJECT'] = True
 
     This configuration variable was added in the [version 1.3.0](/changelog/#version-130).
 
+### APISPEC_DECORATORS
```
[the remark this answers is not in the document]
Comment (FarmerChillax) [reply docs/configuration.md]: I also hesitated between these two, i have chagned to `SPEC_DECORATORS`.

File: src/apiflask/app.py (hunk 2 above)
Comment (FarmerChillax) [reply src/apiflask/app.py]: Yes, it would be better.

**Comment by greyli** [comment]
Merged, thanks!

P.S. I made some improvements in https://github.com/apiflask/apiflask/pull/508/commits/08b48bcedb6a7a3ef76942bcf0954219a4d65d79

**Pull request merged**